receiver/splunkhecreceiver/splunk_to_logdata.go

```diff
@@ -0,0 +1,162 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package splunkhecreceiver
+
+import (
+	"errors"
+	"sort"
+
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/translator/conventions"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/common/splunk"
+)
+
+const (
+	cannotConvertValue = "cannot convert field value to attribute"
+)
+
+// SplunkHecToLogData transforms splunk events into logs
+func SplunkHecToLogData(logger *zap.Logger, events []*splunk.Event) (pdata.ResourceLogsSlice, error) {
+	lrs := pdata.NewResourceLogsSlice()
+	lrs.Resize(len(events))
+
+	for i, event := range events {
+		lr := lrs.At(i)
+		lr.InitEmpty()
+		logRecord := pdata.NewLogRecord()
+		logRecord.InitEmpty()
+
+		// The SourceType field is the most logical "name" of the event.
+		logRecord.SetName(event.SourceType)
+		logRecord.Body().InitEmpty()
+		if value, ok := event.Event.(string); ok {
+			logRecord.Body().SetStringVal(value)
+		} else if value, ok := event.Event.(int64); ok {
+			logRecord.Body().SetIntVal(value)
+		} else if value, ok := event.Event.(float64); ok {
+			logRecord.Body().SetDoubleVal(value)
+		} else if value, ok := event.Event.(bool); ok {
+			logRecord.Body().SetBoolVal(value)
+		} else if value, ok := event.Event.(map[string]interface{}); ok {
+			mapValue, err := convertToAttributeMap(logger, value)
+			if err != nil {
+				return lrs, err
+			}
+			logRecord.Body().SetMapVal(mapValue)
+		} else if value, ok := event.Event.([]interface{}); ok {
+			arrValue, err := convertToArrayVal(logger, value)
+			if err != nil {
+				return lrs, err
+			}
+			logRecord.Body().SetArrayVal(arrValue)
+		}
+
+		// Splunk timestamps are in seconds so convert to nanos by multiplying
+		// by 1 billion.
+		logRecord.SetTimestamp(pdata.TimestampUnixNano(event.Time * 1e9))
+
+		lr.Resource().InitEmpty()
+		attrs := lr.Resource().Attributes()
+		attrs.InitEmptyWithCapacity(3 + len(event.Fields))
```

Comment:
event.Fields is put on the log record attributes so adding `len(event.Fields)` is unnecessary.